validate/validate.go

```diff
@@ -303,50 +309,72 @@ func validateTcb(report *spb.Report, vcekTcb kds.TCBVersion, options *Options) e
 		return fmt.Errorf("chip's VCEK TCB %x does not match the REPORTED_TCB %x",
 			vcekTcb, report.GetReportedTcb())
 	}
-	if !options.PermitProvisionalFirmware {
-		if kds.TCBVersion(report.GetCurrentTcb()) != vcekTcb {
-			return fmt.Errorf("chip's VCEK TCB %x does not match the CURRENT_TCB %x",
-				vcekTcb, report.GetReportedTcb())
-		}
-		if report.GetCurrentTcb() != report.GetCommittedTcb() {
-			return fmt.Errorf("firmware's committed TCB %x does not match the current TCB %x",
-				report.GetCommittedTcb(), report.GetCurrentTcb())
-		}
-	} else if kds.TCBVersion(report.GetCurrentTcb()) < vcekTcb {
-		return fmt.Errorf("firmware's current TCB %x is less than the TCB the VCEK is certified for %x",
-			report.GetCurrentTcb(), vcekTcb)
+	if !kds.TCBPartsLE(vcekTcbParts, currentTcb) {
+		return fmt.Errorf("the TCB the VCEK is certified for %v is not component-wise less than or equal the report's CURRENT_TCB %v",
+			currentTcb, vcekTcbParts)
 	}
-	min, err := kds.ComposeTCBParts(options.MinimumTCB)
-	if err != nil {
-		return fmt.Errorf("option MinimumTCB error: %v", err)
+	if !kds.TCBPartsLE(options.MinimumTCB, vcekTcbParts) {
+		return fmt.Errorf("the policy minimum TCB %v is not component-wise less than or equal the TCB the VCEK is certified for %v",
+			options.MinimumTCB, vcekTcbParts)
 	}
-	if kds.TCBVersion(report.GetCurrentTcb()) < min {
-		return fmt.Errorf("firmware's current TCB %x is less than required %x",
-			report.GetCurrentTcb(), min)
+	return nil
+}
+
+func validateTcb(report *spb.Report, vcekTcb kds.TCBVersion, options *Options) error {
+	reportedTcb := kds.DecomposeTCBVersion(kds.TCBVersion(report.GetReportedTcb()))
+	currentTcb := kds.DecomposeTCBVersion(kds.TCBVersion(report.GetCurrentTcb()))
+	committedTcb := kds.DecomposeTCBVersion(kds.TCBVersion(report.GetCommittedTcb()))
+	launchTcb := kds.DecomposeTCBVersion(kds.TCBVersion(report.GetLaunchTcb()))
```

Comment:
I think we should have some comments to explain all these various TCBs and why we need all these.

Reply:
Reorganized the code to make all the policy decisions more apparent, and added description of each one.